**Раздел 1.5. Задание для самостоятельной работы** (датасет с сайта https://www.kaggle.com/datasets/ruchi798/data-science-job-salaries).
*   Сгруппируйте вакансии по направлениями (DS, DE, Software Engenering, etc.)
*   Какая средняя и медианная зарплата по группам вакансий?
*   Какая средняя и медианная зарплата по каждому региону?
*   Какая самая высокооплачиваемая из групп вакансий, исходя из их средних зарплат?
*   Какое процентное соотношение каждого региона по вакансиям от всех вакансий?
*   Какая корреляция уровня опыта от зарплаты?
*   Сколько должностей в наборе данных?
*   Какие 10 наиболее часто встречающихся должностей? 



In [2]:
import pandas as pd
import numpy as np

In [3]:
df = pd.read_csv('ds_salaries.csv', index_col='Unnamed: 0')

In [4]:
df

work_year experience_level employment_type                   job_title  \
0         2020               MI              FT              Data Scientist   
1         2020               SE              FT  Machine Learning Scientist   
2         2020               SE              FT           Big Data Engineer   
3         2020               MI              FT        Product Data Analyst   
4         2020               SE              FT   Machine Learning Engineer   
..         ...              ...             ...                         ...   
602       2022               SE              FT               Data Engineer   
603       2022               SE              FT               Data Engineer   
604       2022               SE              FT                Data Analyst   
605       2022               SE              FT                Data Analyst   
606       2022               MI              FT                AI Scientist   

     salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0     70000             EUR          79833                 DE             0   
1    260000             USD         260000                 JP             0   
2     85000             GBP         109024                 GB            50   
3     20000             USD          20000                 HN             0   
4    150000             USD         150000                 US            50   
..      ...             ...            ...                ...           ...   
602  154000             USD         154000                 US           100   
603  126000             USD         126000                 US           100   
604  129000             USD         129000                 US             0   
605  150000             USD         150000                 US           100   
606  200000             USD         200000                 IN           100   

    company_location company_size  
0                 DE            L  
1                 JP            S  
2                 GB            M  
3                 HN            S  
4                 US            L  
..               ...          ...  
602               US            M  
603               US            M  
604               US            M  
605               US            M  
606               US            L  

[607 rows x 11 columns]

**Сгруппируйте вакансии по направлениями (DS, DE, Software Engenering, etc.)**

In [5]:
df.groupby('job_title').salary.count()

job_title
3D Computer Vision Researcher                 1
AI Scientist                                  7
Analytics Engineer                            4
Applied Data Scientist                        5
Applied Machine Learning Scientist            4
BI Data Analyst                               6
Big Data Architect                            1
Big Data Engineer                             8
Business Data Analyst                         5
Cloud Data Engineer                           2
Computer Vision Engineer                      6
Computer Vision Software Engineer             3
Data Analyst                                 97
Data Analytics Engineer                       4
Data Analytics Lead                           1
Data Analytics Manager                        7
Data Architect                               11
Data Engineer                               132
Data Engineering Manager                      5
Data Science Consultant                       7
Data Science Engineer         

Видим много разных формулировок вакансий, поэтому создадим дополнительный столбец в DataFrame, в котором будем классифицировать вакансии по ключевым словам.

In [6]:
def job_group_calc(job):
  """Функция, которая определяет принадлежность вакансии к какой-либо группе"""
  if 'Data Scientist' in job or 'Data Science' in job or 'Data Analyst' in job or 'Data Analytics' in job or 'Head of Data' in job:
    return 'DS'
  elif 'Data Engineer' in job or 'Analytics Engineer' in job or 'Data Architect' in job or 'Data Specialist' in job:
    return 'DE'
  elif 'Machine Learning' in job or 'ML ' in job:
    return 'ML'
  elif 'Research Scientist' in job:
    return 'RS'
  elif 'Computer Vision' in job:
    return 'CV'
  elif 'AI Scientist' in job:
    return 'AI'
  else:
    return 'Oth'

In [7]:
df['job_group'] = pd.Series([job_group_calc(job) for job in df.job_title.to_list()]) # создаем новую колонку, где будем сохранять тип вакансии
df

work_year experience_level employment_type                   job_title  \
0         2020               MI              FT              Data Scientist   
1         2020               SE              FT  Machine Learning Scientist   
2         2020               SE              FT           Big Data Engineer   
3         2020               MI              FT        Product Data Analyst   
4         2020               SE              FT   Machine Learning Engineer   
..         ...              ...             ...                         ...   
602       2022               SE              FT               Data Engineer   
603       2022               SE              FT               Data Engineer   
604       2022               SE              FT                Data Analyst   
605       2022               SE              FT                Data Analyst   
606       2022               MI              FT                AI Scientist   

     salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0     70000             EUR          79833                 DE             0   
1    260000             USD         260000                 JP             0   
2     85000             GBP         109024                 GB            50   
3     20000             USD          20000                 HN             0   
4    150000             USD         150000                 US            50   
..      ...             ...            ...                ...           ...   
602  154000             USD         154000                 US           100   
603  126000             USD         126000                 US           100   
604  129000             USD         129000                 US             0   
605  150000             USD         150000                 US           100   
606  200000             USD         200000                 IN           100   

    company_location company_size job_group  
0                 DE            L        DS  
1                 JP            S        ML  
2                 GB            M        DE  
3                 HN            S        DS  
4                 US            L        ML  
..               ...          ...       ...  
602               US            M        DE  
603               US            M        DE  
604               US            M        DS  
605               US            M        DS  
606               US            L        AI  

[607 rows x 12 columns]

Делаем группировку по направлениям

In [8]:
df.job_group.value_counts()

DS     328
DE     175
ML      68
RS      16
CV      10
AI       7
Oth      3
Name: job_group, dtype: int64

Вывод: наибольшее количество вакансий по направлению DS - 328 штук.

**Какая средняя и медианная зарплата по группам вакансий?**

In [9]:
df_group = df.groupby('job_group')['salary_in_usd'].agg(['mean', 'median'])
df_group

mean    median
job_group                         
AI          66135.571429   45896.0
CV          58767.100000   44304.5
DE         121251.777143  112872.0
DS         110725.265244  102919.5
ML         113026.132353   87932.0
Oth         49050.000000   54957.0
RS         109019.500000   76263.5

Вывод: найдены средняя и медианная зарплата по группам вакансий

**Какая средняя и медианная зарплата по каждому региону?**

In [10]:
df.company_location.unique()

array(['DE', 'JP', 'GB', 'HN', 'US', 'HU', 'NZ', 'FR', 'IN', 'PK', 'CN',
       'GR', 'AE', 'NL', 'MX', 'CA', 'AT', 'NG', 'ES', 'PT', 'DK', 'IT',
       'HR', 'LU', 'PL', 'SG', 'RO', 'IQ', 'BR', 'BE', 'UA', 'IL', 'RU',
       'MT', 'CL', 'IR', 'CO', 'MD', 'KE', 'SI', 'CH', 'VN', 'AS', 'TR',
       'CZ', 'DZ', 'EE', 'MY', 'AU', 'IE'], dtype=object)

Видим много разных стран. Сгруппируем страны по регионам и создадим новую колонку в DataFrame.

In [11]:
def region_calc(country):
  """Функция, определяющая регион страны по ее обозначению"""
  if country in ['DE', 'GB', 'HU', 'FR', 'GR', 'NL', 'AT', 'ES', 'PT', 'DK', 'IT', 'HR', 'LU', 'PL', 'RO', 'BE', 'UA', 'RU', 'MT', 'MD', 'SI', 'CH', 'CZ', 'EE', 'IE']:
    return 'Europe'
  elif country in ['HN', 'US', 'MX', 'CA']:
    return 'N.America'
  elif country in ['BR', 'CL', 'CO']:
    return 'S.America'
  elif country in ['JP', 'IN', 'PK', 'CN', 'AE', 'SG', 'IQ', 'IL', 'IR', 'VN', 'TR', 'MY']:
    return 'Asia'
  elif country in ['NZ', 'AS', 'AU']:
    return 'Australia'
  elif country in ['NG', 'KE', 'DZ']:
    return 'Africa'
  else:
    return 'Other'

In [12]:
df['region'] = pd.Series([region_calc(country) for country in df.company_location.to_list()]) # создаем новую колонку, где будем сохранять регион
df

work_year experience_level employment_type                   job_title  \
0         2020               MI              FT              Data Scientist   
1         2020               SE              FT  Machine Learning Scientist   
2         2020               SE              FT           Big Data Engineer   
3         2020               MI              FT        Product Data Analyst   
4         2020               SE              FT   Machine Learning Engineer   
..         ...              ...             ...                         ...   
602       2022               SE              FT               Data Engineer   
603       2022               SE              FT               Data Engineer   
604       2022               SE              FT                Data Analyst   
605       2022               SE              FT                Data Analyst   
606       2022               MI              FT                AI Scientist   

     salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0     70000             EUR          79833                 DE             0   
1    260000             USD         260000                 JP             0   
2     85000             GBP         109024                 GB            50   
3     20000             USD          20000                 HN             0   
4    150000             USD         150000                 US            50   
..      ...             ...            ...                ...           ...   
602  154000             USD         154000                 US           100   
603  126000             USD         126000                 US           100   
604  129000             USD         129000                 US             0   
605  150000             USD         150000                 US           100   
606  200000             USD         200000                 IN           100   

    company_location company_size job_group     region  
0                 DE            L        DS     Europe  
1                 JP            S        ML       Asia  
2                 GB            M        DE     Europe  
3                 HN            S        DS  N.America  
4                 US            L        ML  N.America  
..               ...          ...       ...        ...  
602               US            M        DE  N.America  
603               US            M        DE  N.America  
604               US            M        DS  N.America  
605               US            M        DS  N.America  
606               US            L        AI  N.America  

[607 rows x 13 columns]

Находим среднюю и медианную зарплату по регионам

In [13]:
df_region = df.groupby('region')['salary_in_usd'].agg(['mean', 'median'])
df_region

mean    median
region                            
Africa      42318.000000   30000.0
Asia        48312.765957   29751.0
Australia   93436.200000   87425.0
Europe      69358.369427   63831.0
N.America  139461.953728  130000.0
S.America   23538.000000   21844.0

Вывод: найдены средняя и медианная зарплата по региону расположения компании

**Какая самая высокооплачиваемая из групп вакансий, исходя из их средних зарплат?**

In [14]:
max_salary_group = df_group[df_group['mean']==df_group['mean'].max()]
max_salary_group

mean    median
job_group                         
DE         121251.777143  112872.0

Вывод: самая высокооплачиваемая группа вакансий - DE.

**Какое процентное соотношение каждого региона по вакансиям от всех вакансий?**

In [15]:
df_region_count = df.groupby('region').region.count()
df_region_percent = round(100 * df_region_count / df_region_count.sum(), 2)
df_region_percent

region
Africa        0.66
Asia          7.74
Australia     0.82
Europe       25.86
N.America    64.09
S.America     0.82
Name: region, dtype: float64

Вывод: наибольший процент от общего количества вакансий в Северной Америке.

**Какая корреляция уровня опыта от зарплаты?**

Создаем вспомогательную таблицу с двумя столбцами

In [16]:
df_exp = pd.DataFrame(df, columns=['experience_level', 'salary_in_usd'])
df_exp.experience_level.unique()
# EN Entry-level / Junior MI Mid-level / Intermediate SE Senior-level / Expert EX Executive-level 

array(['MI', 'SE', 'EN', 'EX'], dtype=object)

In [17]:
def rank_exp(exp_level):
  """Ранжирование уровней опыта"""
  if exp_level == 'EN':
    return 1
  elif exp_level == 'MI':
    return 2
  elif exp_level == 'SE':
    return 3
  elif exp_level == 'EX':
    return 4
  else:
    return None

In [18]:
df_exp['exp_lvl'] = pd.Series([rank_exp(exp_level) for exp_level in df_exp.experience_level.to_list()]) # создаем новую колонку, где будем сохранять ранжированный опыт
df_exp

experience_level  salary_in_usd  exp_lvl
0                 MI          79833        2
1                 SE         260000        3
2                 SE         109024        3
3                 MI          20000        2
4                 SE         150000        3
..               ...            ...      ...
602               SE         154000        3
603               SE         126000        3
604               SE         129000        3
605               SE         150000        3
606               MI         200000        2

[607 rows x 3 columns]

In [19]:
df_exp.corr(method='spearman')

salary_in_usd   exp_lvl
salary_in_usd       1.000000  0.558567
exp_lvl             0.558567  1.000000

Коэффициент ранговой корреляции Спирмена между уровнем опыта и зарплатой = 0,56

**Сколько должностей в наборе данных?**

In [20]:
df.groupby('job_title').job_title.unique().count()

50

**Какие 10 наиболее часто встречающихся должностей?**

In [23]:
df.groupby('job_title').job_title.count().nlargest(n=10)

job_title
Data Scientist                143
Data Engineer                 132
Data Analyst                   97
Machine Learning Engineer      41
Research Scientist             16
Data Science Manager           12
Data Architect                 11
Big Data Engineer               8
Machine Learning Scientist      8
AI Scientist                    7
Name: job_title, dtype: int64

Вывод: получены 10 наиболее часто встречающихся должностей.